pkg/ottl/ottlfuncs/func_parse_key_value.go

```diff
@@ -0,0 +1,121 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+type ParseKeyValueArguments[K any] struct {
+	Target        ottl.StringGetter[K]
+	Delimiter     ottl.Optional[string]
+	PairDelimiter ottl.Optional[string]
```

Comment:
Can you add tests for what happens if either of these are an empty string? Also what happens if they are longer than the target, or aren't found in the target?

Reply:
Good catch, the current implementation allows either to be equal to `""` which should not be allowed. In the case one of the delimiters is set to an empty string, should the function error or use the default?

Reply:
I could easily be misunderstanding something, but trying the following test case (`expected` doesn't matter here):
```golang
               {
			name: "empty delimiters",
			target: ottl.StandardStringGetter[any]{
				Getter: func(ctx context.Context, tCtx any) (any, error) {
					return "a=b c=d", nil
				},
			},
			delimiter:      ottl.NewTestingOptional(""),
			pair_delimiter: ottl.NewTestingOptional(""),
			expected:       map[string]any{},
		},
```
Errors on the check for equal delimiters in the start of `parseKeyValue()`:
`Error: Received unexpected error: pair delimiter "" cannot be equal to delimiter ""`

Which tells me it is possible for the delimiter variables to get set with the empty string if the user passes them in. Obviously in this case they can't both be set because it gets caught, but it demonstrates that one of them can be set to `""`.

Reply:
Unless you're referring to when the OTTL function would be used while the collector is running and not isolated like in a unit test?

Reply:
I'm referring to what the user may specify in their config. Whether they omit either parameter or explicitly set it to `""`, it should change the delimiter to the default value.

Reply:
Ahhh, yep, I misunderstood your first comment. That behavior works for me and I can change it to that.

Reply:
>Whether they omit either parameter or explicitly set it to `""`, it should change the delimiter to the default value.

We can check if the user passed a parameter using the `IsEmpty()` function on the `Optional` type. In my opinion, if we want to forbid these values, we should check if they have been explicitly set to an invalid value and throw an error instead of using the default, or at a minimum print a warning. I would be surprised if the value I set was silently ignored by the function.

Reply:
I'll have it throw an error since I don't really see an example of logging a warning in another OTTL function.